File: src/main/java/math/LinearInterpolation.java
package math;

import java.util.Arrays;

public class LinearInterpolation {

    public static Double[] interpolate(Double[] x, Double[] y, Double[] xi) throws IllegalArgumentException {

        if (x.length != y.length) {
            throw new IllegalArgumentException("X and Y must be the same length");
        }
        if (x.length == 1) {
            throw new IllegalArgumentException("X must contain more than one value");
        }
        double[] dx = new double[x.length - 1];
        double[] dy = new double[x.length - 1];
        double[] slope = new double[x.length - 1];
        double[] intercept = new double[x.length - 1];

        // Calculate the line equation (i.e. slope and intercept) between each point
        for (int i = 0; i < x.length - 1; i++) {
            dx[i] = x[i + 1] - x[i];
            if (dx[i] == 0) {
                x[i + 1] += x[i + 1] * 0.01;
            }
            if (dx[i] < 0) {
                System.out.println(Arrays.toString(x));
                throw new IllegalArgumentException("X must be sorted " + x[i + 1] + " " + x[i]);
            }

            dy[i] = y[i + 1] - y[i];
            slope[i] = dy[i] / dx[i];
            intercept[i] = y[i] - x[i] * slope[i];
        }

        // Perform the interpolation here
        Double[] yi = new Double[xi.length];
        for (int i = 0; i < xi.length; i++) {
            if ((xi[i] < x[0])) {
                yi[i] = Double.NaN;
            } else {
                int loc = Arrays.binarySearch(x, xi[i]);

                if (loc < -1) {
                    loc = -loc - 2;
                    yi[i] = Math.min(slope[Math.min(loc, slope.length -1)] * xi[i] + intercept[Math.min(loc, intercept.length -1)], y[y.length -1]);
                }
                else {
                    yi[i] = y[loc];
                }
            }
        }

        return yi;
    }
}

File: src/main/kotlin/domain/math/LinearInterpolation.kt
package domain.math

import java.util.Arrays

object LinearInterpolation {

    fun interpolate(xIn: Array<Double>, yIn: Array<Double>, xi: Array<Double>): Array<Double> {
        val x = xIn.clone()
        val y = yIn.clone()

        require(x.size == y.size) { "X and Y must be the same length" }
        require(x.size > 1) { "X must contain more than one value" }

        val dx = DoubleArray(x.size - 1)
        val dy = DoubleArray(x.size - 1)
        val slope = DoubleArray(x.size - 1)
        val intercept = DoubleArray(x.size - 1)

        for (i in 0 until x.size - 1) {
            dx[i] = x[i + 1] - x[i]
            if (dx[i] < 0) {
                throw IllegalArgumentException("X must be sorted ${x[i + 1]} ${x[i]}")
            }
            if (dx[i] == 0.0) {
                x[i + 1] += x[i + 1] * 0.01
            }
            dy[i] = y[i + 1] - y[i]
            slope[i] = dy[i] / dx[i]
            intercept[i] = y[i] - x[i] * slope[i]
        }

        return Array(xi.size) { i ->
            if (xi[i] < x[0]) {
                0.0
            } else {
                var loc = Arrays.binarySearch(x, xi[i])
                if (loc < -1) {
                    loc = -loc - 2
                    minOf(
                        slope[minOf(loc, slope.size - 1)] * xi[i] + intercept[minOf(loc, intercept.size - 1)],
                        y[y.size - 1]
                    )
                } else {
                    y[loc]
                }
            }
        }
    }
}
